server/quota.go

```diff
@@ -15,86 +15,73 @@
 package server
 
 import (
-	"database/sql"
 	"fmt"
+	"sync"
 
-	"github.com/coreos/etcd/clientv3"
-	"github.com/golang/glog"
 	"github.com/google/trillian/quota"
-	"github.com/google/trillian/quota/cacheqm"
-	"github.com/google/trillian/quota/etcd/etcdqm"
-	"github.com/google/trillian/quota/mysqlqm"
 )
 
 const (
 	// QuotaNoop represents the noop quota implementation.
 	QuotaNoop = "noop"
+)
 
-	// QuotaMySQL represents the MySQL quota implementation.
-	QuotaMySQL = "mysql"
+// NewQuotaManagerFunc is the signature of a function which can be registered
+// to provide instances of a quota manager.
+type NewQuotaManagerFunc func() (quota.Manager, error)
 
-	// QuotaEtcd represents the etcd quota implementation.
-	QuotaEtcd = "etcd"
+var (
+	qpMu     sync.RWMutex
+	qpOnce   sync.Once
+	qpByName map[string]NewQuotaManagerFunc
 )
 
-// QuotaParams represents all parameters required to initialize a quota.Manager.
-//
-// Depending on the supplied QuotaSystem, the actual quota.Manager implementation, as returned by
-// NewQuotaManager, may differ.
-//
-// See fields for details.
-type QuotaParams struct {
-	// QuotaSystem represents the underlying quota implementation used.
-	// Valid values are "noop", "mysql" and "etcd".
-	QuotaSystem string
-
-	// DB is the database used by MySQL quotas.
-	DB *sql.DB
+func init() {
+	if err := RegisterQuotaManager(QuotaNoop, func() (quota.Manager, error) {
+		return quota.Noop(), nil
+	}); err != nil {
+		panic(err)
```

Comment:
Again, not sure panic is the right thing.

Reply:
glogd